### PR TITLE
ENH: infer Timestamp(iso8601string) resolution

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.0.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:340]:
```suggestion
- Passing a string in ISO-8601 format to :class:`Timestamp` will retain the resolution of the parsed input if it is "s", "ms", "us", or "ns"; otherwise it will be cast to the closest supported resolution (:issue:`49737`)
```

File: pandas/core/arrays/datetimes.py
Comment (mroeschke) [inline pandas/core/arrays/datetimes.py:517]: Does this mean `as_unit` will need to be called eventually?

File: pandas/tests/scalar/timestamp/test_unary_ops.py
Comment (mroeschke) [inline pandas/tests/scalar/timestamp/test_unary_ops.py:261]: Still needed?

Comment (jbrockmendel) [reply pandas/tests/scalar/timestamp/test_unary_ops.py]: nope, removed

File: pandas/core/arrays/datetimes.py
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: nope, removed

File: pandas/core/computation/pytables.py
Comment (MarcoGorelli) [inline pandas/core/computation/pytables.py:220]: Is this thanks to #48819, which introduced `NaT.as_unit`?

**Review by MarcoGorelli — COMMENTED** [review]
From a first pass looks good, there's just a couple of merge conflicts to solve

File: pandas/core/computation/pytables.py
Comment (jbrockmendel) [reply pandas/core/computation/pytables.py:220]: correct

**Review by mroeschke — COMMENTED** [review]
Looks pretty good. Do we have any tests where the iso string has a UTC offset?

**Comment by jbrockmendel** [comment]
> Looks pretty good. Do we have any tests where the iso string has a UTC offset?

im pretty sure we have loads of these, unless we're thinking of different things.  you mean like "2016-01-01 04:05:06-01:00" right?

**Comment by mroeschke** [comment]
> > Looks pretty good. Do we have any tests where the iso string has a UTC offset?
> 
> im pretty sure we have loads of these, unless we're thinking of different things. you mean like "2016-01-01 04:05:06-01:00" right?

Yeah exactly. Should those test assert that the resolution is a certain value too?

**Review by mroeschke — APPROVED** [review]
LGTM (I think the failure are unrelated) Merge when ready @MarcoGorelli

File: pandas/tests/tseries/offsets/test_common.py
Comment (MarcoGorelli) [inline pandas/tests/tseries/offsets/test_common.py:155]: I'm confused -  the xfail reason only mentions `t.tzinfo`, and the if-statement condition depends on `_offset`, yet `t = Timestamp("20080101", tz=tz)` doesn't depend on `_offset`, why's that?

**Review by MarcoGorelli — COMMENTED** [review]
Looks good, I've just left some questions I'd like to double-check before merging myself - sorry if they're basic, I'm just a bit concerned about the bus factor on some of this work

Not blockers though, feel free to merge without my approval here

File: pandas/_libs/tslibs/conversion.pyx
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:541]: some combination of that and me not wanting to recurse.  honestly i dont remember how much of each off the top of my head

File: pandas/tests/tseries/offsets/test_common.py
Comment (jbrockmendel) [reply pandas/tests/tseries/offsets/test_common.py:155]: best guess is some kind of windows-themed demon trying to increase my blood pressure.  seriously i couldn't reproduce it so just punted.

Comment (MarcoGorelli) [reply pandas/tests/tseries/offsets/test_common.py:155]: OK, no objections. I don't think I've ever seen anyone use `tz_local` anyway

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me pending green (I may not be awake by the time CI finishes, but no objections to this being merged)

**Comment by MarcoGorelli** [comment]
looks good - merging then, thanks @jbrockmendel !

**Pull request merged**
